kernel/src/engine/arrow_utils.rs

```diff
@@ -1060,13 +1062,34 @@ fn parse_json_impl(json_strings: &StringArray, schema: ArrowSchemaRef) -> DeltaR
 // TODO (zach): this should stream data to the JSON writer and output an iterator.
 #[internal_api]
 pub(crate) fn to_json_bytes(
-    data: impl Iterator<Item = DeltaResult<Box<dyn EngineData>>> + Send,
+    data: impl Iterator<Item = DeltaResult<FilteredEngineData>> + Send,
 ) -> DeltaResult<Vec<u8>> {
     let mut writer = LineDelimitedWriter::new(Vec::new());
     for chunk in data {
-        let arrow_data = ArrowEngineData::try_from_engine_data(chunk?)?;
-        let record_batch = arrow_data.record_batch();
-        writer.write(record_batch)?;
+        let filtered_data = chunk?;
+        // Extract the underlying data and apply the selection vector to get only selected rows
+        let batch = extract_record_batch(&*filtered_data.data)?;
+
+        // Honor the new contract: if selection vector is shorter than the number of rows,
+        // then all rows not covered by the selection vector are assumed to be selected
+        let num_rows = batch.num_rows();
+        let mut selection_vector = filtered_data.selection_vector.clone();
+
+        if selection_vector.is_empty() {
+            // If selection vector is empty, write all rows per contract.
+            writer.write(batch)?;
+        } else {
+            if selection_vector.len() < num_rows {
+                // Extend the selection vector with `true` for uncovered rows
+                selection_vector.resize(num_rows, true);
+            } else if selection_vector.len() > num_rows {
+                // Take a sublist of the selection vector equal to the data size
+                selection_vector = selection_vector[..num_rows].to_vec();
```

Comment:
This indicates a bad FilteredEngineData, so I prefer we return an error here. 
```suggestion
            } else if selection_vector.len() > num_rows {
                return Err(Error::generic("Selection vectors must have fewer or equal rows to data in FilteredEngineData. Data had {num_rows} rows, while selection vector had {selection_vector.len()} rows."))
```